odoo/index.py

```diff
@@ -0,0 +1,285 @@
+    'inner',
+    'intersect',
+    'into',
+    'is',
+    'isnull',
+    'join',
+    'lateral',
+    'leading',
+    'left',
+    'like',
+    'limit',
+    'localtime',
+    'localtimestamp',
+    'natural',
+    'not',
+    'notnull',
+    'null',
+    'offset',
+    'on',
+    'only',
+    'or',
+    'order',
+    'outer',
+    'overlaps',
+    'placing',
+    'primary',
+    'references',
+    'returning',
+    'right',
+    'select',
+    'session_user',
+    'similar',
+    'some',
+    'symmetric',
+    'system_user',
+    'table',
+    'tablesample',
+    'then',
+    'to',
+    'trailing',
+    'true',
+    'union',
+    'unique',
+    'user',
+    'using',
+    'variadic',
+    'verbose',
+    'when',
+    'where',
+    'window',
+    'with',
+}
+
+
+def maybe_quote(ident: str) -> str:
+    # SQL identifiers and key words must begin with a letter (a-z, but also
+    # letters with diacritical marks and non-Latin letters) or an underscore
+    # (_). Subsequent characters in an identifier or key word can be letters,
+    # underscores, digits (0-9), or dollar signs ($).
+    # FIXME: find stricter SQL definition but this will do for now (?)
+    # odoo is generally case-sensitive so non-lowercase names must be quoted
+    if ident.isidentifier() and ident not in RESERVED and ident.casefold() == ident:
+        return ident
+    return quote(ident)
+
+def install(env, field, index: 'index', existing: Optional[str] = None):
+    expr = index.serialize(env, field)
+    if not expr:
+        return
+
+    if existing:
+        if expr.lower() == existing.lower():
+            return
+
+        _logger.warning(
+            "Ignoring index: already exists with a different definition, "
+            "you may want to update the index to the new version\n"
+            "current: %s\n    new: %s",
+            existing,
+            expr
+        )
+        return
+
+    with env.cr.savepoint(flush=False):
+        env.cr.execute(expr)
+
+class Nulls(enum.Enum):
+    # maybe as convenient alias for $field IS NOT NULL?
+    # NO = enum.auto()
+    DISTINCT = enum.auto()
+    # PG15 only
+    # NOT_DISTINCT = enum.auto()
+
+@dataclasses.dataclass
+class index(ABC):
+    # partial index condition, `{field}` will expand to the host field's name
+    # (for single-field indexes)
+    where: str = ''
+
+    method: str = 'btree'
+
+    def serialize(self, env, field):
+        expression = self.expression(env, field)
+        if expression is None:
+            return
+
+        schema = 'public'
+        table = env[field.model_name]._table
+        # TODO: pass table and name from outside?
+        index_name = sql.make_index_name(table, field.name)
+        where = ''
+        if self.where:
+            if self.where.format(field=field.name).isidentifier():
+                where = f' where {self.where}'
+            else:
+                where = f' where ({self.where})'
+        return (
+            f'create index {maybe_quote(index_name)} '
```

Comment:
I generally try to use uppercase for SQL keywords, because it makes SQL debugging easier (everything else is generally lowercase).